docker/tarfile/src.go

```diff
@@ -43,28 +42,20 @@ type layerInfo struct {
 //       the source of an image.
 // 	To do for both the NewSourceFromFile and NewSourceFromStream functions
 
-// NewSourceFromFile returns a tarfile.Source for the specified path
-// NewSourceFromFile supports both conpressed and uncompressed input
+// NewSourceFromFile returns a tarfile.Source for the specified path.
 func NewSourceFromFile(path string) (*Source, error) {
 	file, err := os.Open(path)
 	if err != nil {
 		return nil, errors.Wrapf(err, "error opening file %q", path)
 	}
 	defer file.Close()
 
-	reader, err := gzip.NewReader(file)
-	if err != nil {
-		return &Source{
-			tarPath: path,
-		}, nil
```

Comment:
(Just a quick question for now:) Is there an user who benefits from adding decompression support for streams? The only user of `NewSourceFromStream` in c/image processes the output of `docker save`, which is not compressed.

If there is no user, I’d rather leave the decompression exclusive to `NewSourceFromFile`, or rewrite the code in another way to preserve the property that `NewSourceFromFile` on an uncompressed file doesn’t make an unnecessary (and pretty costly) copy.

Reply:
If you want, I can rework this so that it just uses the original file if it is uncompressed (I actually missed that the code did this when I deleted this hunk -- oops!).

But I don't see the negative of adding it to the stream implementation. The uncompressed path for streams is basically identical to the compressed path (minus the overhead of reading 5 bytes rather than using `bufio` for the entire read -- which isn't a significant overhead. So adding it to the stream implementation just means that `containers/image` users can stream compressed archives directly to c/image rather than making two copies.

Reply:
Fixed

Reply:
> So adding it to the stream implementation just means that containers/image users can stream compressed archives directly to c/image rather than making two copies.

This does not actually change that, because `docker-archive:` always calls `NewSourceFromFile` (`docker/tarball` is an internal helper, not a transport with its own `ImageReference` that could be used via `copy.Image`).  Which is why was asking about an user who benefits.

Sure, this does not hurt, and the consistency in the API is nice :)

Reply:
I was thinking more about users who might want to use `docker/tarball` but there are probably no such users. Anyway, I've fixed up the code to now no longer make a copy if the archive is already uncompressed.